naga/src/front/atomic_upgrade.rs

```diff
@@ -0,0 +1,373 @@
+        padding.debug("upgrading expr: ", handle);
+        let expr = if let Some(fh) = maybe_fn_handle {
+            let function = self.module.functions.try_get(fh)?;
+            function.expressions.try_get(handle)?.clone()
+        } else {
+            self.module.global_expressions.try_get(handle)?.clone()
+        };
+
+        padding.debug("expr: ", &expr);
+        let new_expr = match expr.clone() {
+            Expression::AccessIndex { base, index } => Expression::AccessIndex {
+                base: self.upgrade_expression(maybe_fn_handle, base)?,
+                index,
+            },
+            Expression::GlobalVariable(var) => {
+                Expression::GlobalVariable(self.upgrade_global_variable(var)?)
+            }
+            Expression::LocalVariable(var) => Expression::LocalVariable(
+                self.upgrade_local_variable(maybe_fn_handle.ok_or(Error::NoFunction)?, var)?,
+            ),
+            _ => {
+                return Err(Error::Unsupported);
+            }
+        };
+
+        if new_expr != expr {
+            padding.debug("inserting new expr: ", &new_expr);
+            let arena = if let Some(fh) = maybe_fn_handle {
+                let f = self.module.functions.get_mut(fh);
+                &mut f.expressions
+            } else {
+                &mut self.module.global_expressions
+            };
+            let span = arena.get_span(handle);
+            let new_handle = arena.append(new_expr, span);
+            padding.debug("new expr handle: ", new_handle);
+            Ok(new_handle)
+        } else {
+            Ok(handle)
+        }
+    }
+}
+
+impl Module {
+    /// Upgrade all atomics given.
+    pub(crate) fn upgrade_atomics(
+        &mut self,
+        ops: impl IntoIterator<Item = AtomicOp>,
+    ) -> Result<(), Error> {
+        let mut state = UpgradeState {
+            padding: Default::default(),
+            module: self,
+        };
+
+        for op in ops.into_iter() {
+            let padding = state.inc_padding();
+            padding.debug("op: ", op);
+
+            // Find the expression's enclosing function, if any
+            let mut maybe_fn_handle = None;
+            for (fn_handle, function) in state.module.functions.iter() {
+                log::trace!("function: {fn_handle:?}");
+                if function.expressions.try_get(op.pointer_handle).is_ok() {
+                    log::trace!("  is op's function");
+                    maybe_fn_handle = Some(fn_handle);
+                    break;
+                }
+            }
+
+            padding.debug("upgrading the pointer type:", op.pointer_type_handle);
+            let _new_pointer_type_handle = state.upgrade_type(op.pointer_type_handle)?;
+
+            padding.debug("upgrading the pointer expression", op.pointer_handle);
+            let _new_pointer_handle =
+                state.upgrade_expression(maybe_fn_handle, op.pointer_handle)?;
+        }
+
```

Comment:
The next step is to "spiral out" with a search/replace of expressions that use `op.pointer_handle`.

Reply:
(and "upgrade" those to use `_new_pointer_handle`)